pkg/spanconfig/spanconfigsqlwatcher/sql_watcher.go

```diff
@@ -0,0 +1,453 @@
+		stopper:          stopper,
+		knobs:            knobs,
+	}
+}
+
+// Watch is the entry point for the SQLWatcher.
+func (s *SQLWatcher) Watch(ctx context.Context) (<-chan spanconfig.Update, error) {
+	updatesCh := make(chan spanconfig.Update)
+
+	descUpdatesCh, err := s.watchForDescriptorUpdates(ctx)
+	zonesUpdateCh, err := s.watchForZoneConfigUpdates(ctx)
+	if err != nil {
+		log.Warningf(ctx, "error establishing rangefeed over system.descritpors %v", err)
+		return nil, err
+	}
+	if err := s.stopper.RunAsyncTask(ctx, "span-config-reconciliation", func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-s.stopper.ShouldQuiesce():
+				return
+			case descID := <-descUpdatesCh:
+				err := s.onDescIDUpdate(ctx, descID, updatesCh)
+				if err != nil {
+					log.Errorf(ctx, "could not react to desc id %d update err: %v", descID, err)
+				}
+			case descID := <-zonesUpdateCh:
+				err := s.onDescIDUpdate(ctx, descID, updatesCh)
+				if err != nil {
+					log.Errorf(ctx, "could not react to zone config update with id %d err: %v", descID, err)
+				}
+			}
+		}
+	}); err != nil {
+		return nil, err
+	}
+	return updatesCh, nil
+}
+
+// watchForDescriptorUpdates establishes a rangefeed over system.descriptors and
+// sends updates on the returned channel. The rangefeed performs an initial scan
+// over the table unless indicated otherwise by the
+// SQLWatcherDisableInitialScan testing knob.
+func (s *SQLWatcher) watchForDescriptorUpdates(ctx context.Context) (<-chan descpb.ID, error) {
+	updatesCh := make(chan descpb.ID)
+	descriptorTableStart := s.codec.TablePrefix(keys.DescriptorTableID)
+	descriptorTableSpan := roachpb.Span{
+		Key:    descriptorTableStart,
+		EndKey: descriptorTableStart.PrefixEnd(),
+	}
+	handleEvent := func(ctx context.Context, ev *roachpb.RangeFeedValue) {
+		value := ev.Value
+		if !ev.Value.IsPresent() {
+			// The descriptor was deleted.
+			value = ev.PrevValue
+		}
+		if !value.IsPresent() {
+			return
+		}
+
+		var descriptor descpb.Descriptor
+		if err := value.GetProto(&descriptor); err != nil {
+			logcrash.ReportOrPanic(
+				ctx,
+				&s.settings.SV,
+				"%s: failed to unmarshal descriptor %v",
+				ev.Key,
+				value,
+			)
+		}
+		if descriptor.Union == nil {
+			return
+		}
+
+		table, database, _, _ := descpb.FromDescriptorWithMVCCTimestamp(&descriptor, ev.Value.Timestamp)
+
+		var id descpb.ID
+		if table != nil {
+			id = table.GetID()
+		} else if database != nil {
+			id = database.GetID()
+		} else {
+			// We only care about database or table descriptors being updated.
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+		case updatesCh <- id:
+		}
+	}
+	opts := []rangefeed.Option{
+		rangefeed.WithDiff(),
+	}
+	if !s.knobs.SQLWatcherDisableInitialScan {
+		opts = append(opts, rangefeed.WithInitialScan(nil))
+		opts = append(opts, rangefeed.WithOnInitialScanError(func(ctx context.Context, err error) bool {
+			logcrash.ReportOrPanic(
```

Comment:
Why logcrash?  Should we do it for KVWatcher's initial scan error too? I think not, the settingwatcher seems to expect grpc errors.